### PR TITLE
obs-webrtc: Add user-agent, timestamp, and ssrc generation

### PR DESCRIPTION
**Pull request by mondain**
### Description

The following changes were made to provide compatibility and increase usability behind NAT / Firewall. Specific modifications and additions are as follows:

- Creates a User-Agent string for application identification
- Implements a maximum fragment size for video packetization
- Uses a randomized u32 generation routine for SSRC and RTP timestamps to fit RFC-3550
- Adds additional life-cycle methods for clearer flow; Init, Setup...

### Motivation and Context

My motivation is to remove hard-coded SSRC values and to set starting RTP timestamps to match RFC-3550.

### How Has This Been Tested?

All testing during development was performed on Windows 11 x64 against a modified private pre-release version of Red5 Pro Server. As of generation of this PR, I cannot build on this branch in current windows environment and I'm not sure why; its an error in the PowerShell extension within VS Code that does not appear in my separate fork.

### Types of changes

- New feature
- Tweak
- Code cleanup

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Review by gxalpha — CHANGES_REQUESTED** [review]
Please squash the clang-format commit with your code changes and fix the commit titles in accordance with the [commit guidelines](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst#commit-guidelines).

**Comment by mondain** [comment]
@gxalpha ok will do, whats the process for that, recreation of the PR itself?

File: .gitignore
Comment (tytan652) [inline .gitignore:41]: I may misunderstand something, but workspace are normally outside of the local repo because they agglomerate various folder.

I don't see the point of adding it to the `.gitignore`.

**Comment by gxalpha** [comment]
No, keep the PR open and force-push to the branch you created it from. That will update it.

File: .gitignore
Comment (mondain) [reply .gitignore:41]: Developer preference catch-all; I keep workspaces self-contained, but I'm a new submitter here, I'll remove it from the ignore.

**Comment by pkviet** [comment]
Hello, thanks again for this PR. 👍
There are currently 6 commits, the 1st of which is unnecessary, though i guess it's handy to keep it locally for your setup.
So you could run:
`git rebase -i HEAD~6`
This will open your editor and you'll get a screen like this:
notepad++_2023-06-23_08-19-12
Change to the following:
notepad++_2023-06-23_08-13-42

This will drop the 1st commit, squash all the other commits and retain the title of the 4th commit (notice the f -C).
Save your editor and exit. The rebase will complete.
Then you would run:
`git commit --amend`

and add the description you've written on github, so I assume that'd be:
`The following changes were made to provide compatibility and increase 
usability behind NAT / Firewall. Specific modifications and additions 
are as follows:

Creates a User-Agent string for application identification
Implements a maximum fragment size for video packetization
Uses a randomized u32 generation routine for SSRC and RTP timestamps 
to fit RFC-3550
Adds OPTIONS request support for collecting configuration data;
specifically for this use-case is the Link header containing a 
TURN/STUN server (only a single ICE server is used in this initial 
implementation)
Adds additional life-cycle methods for clearer flow; Init, Options, 
Setup...`
As you can see I've formatted the description to 72 chars max per line.
(That's the guideline only for commits, not on the PR github page itself.)

Also for future reference, the commit prefixes in the titles are per project or module (ex: libobs, UI, obs-webrtc, obs-ffmpeg, etc ...). The 50 chars limitation is for the title, excluding the prefix.
In general, one would split the PR in several commits if they deal with different modules (say, one commit in libobs, another in obs-webrtc etc...) or if there are different features for which one wants to facilitate later debugging (ex: obs-webrtc: Feature 1; obs-webrtc: Feature 2).

Lastly, you force push to your fork which will auto update the PR on github.

**Comment by pkviet** [comment]
@Sean-Der @tt2468 could you review this PR ? thanks.

File: plugins/obs-webrtc/whip-output.cpp
Comment (pkviet) [inline plugins/obs-webrtc/whip-output.cpp:449]: there's an error here thrown in macOS clang because h is unused (warnings are turned into errors in obs).
So either just use the function without defining h if you're sure it can't fail, or if it's not a CURLHE_OK , do a cleanup.

Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:449]: That's in the next push.

**Comment by mondain** [comment]
@pkviet I'm very clearly messing up this chain trying to meet these expectations. I don't know how to fix it.

**Comment by pkviet** [comment]
> @pkviet I'm very clearly messing up this chain trying to meet these expectations. I don't know how to fix it.

No worry, Sean tt2468 and i will take a look.

**Comment by mondain** [comment]
@pkviet okay got the short and long desc in; I'll see if I can squash and roll up now.

**Comment by pkviet** [comment]
Mmh ci is failing for mac. But the runner has a recent enough version so it's odd.
@RytoEX tells me that the runner has libcurl  8.1.2.
7.84 is min required.

**Comment by mondain** [comment]
@pkviet my M1 is failing / complaining about libcurl as well

**Comment by pkviet** [comment]
> @pkviet my M1 is failing / complaining about libcurl as well

I don't have a mac but @PatTheMav checked with 
https://github.com/curl/curl/blob/master/docs/examples/headerapi.c
And it was compiled fine on his mac.

Not sure what's going on there.

**Comment by pkviet** [comment]
can you check on your mac the curl version called in code with curl_version() ? seems curl_easy_header was added in april 22 with 7.83 (see https://curl.se/changes.html ); the CURLHcode errors date back from 8.0 march 23 (but you've removed them)

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:173]: Is there a case that `_WIN64` is defined but `_WIN32` is not defined? Unless such conditions met, `Windows x86_64` won't appear.

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:169]: Do we need to use `std::stringstream` instead of just `const char *` or `std::string`?

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:194]: Maybe `obs_get_version_string()` is ok.
```suggestion
	ua << "(OBS-Studio/" << obs_get_version_string() << "; ";
```

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:175]: Did you intentionally use a lower case `s`?

File: plugins/obs-webrtc/whip-output.h
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.h:27]: I think `constexpr` would be better.
I'm afraid `the` might be redundant in `with the lower the value`  but I'd like to leave it to be commented by another one.

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:19]: Do we need to have this variable in the global scope? It looks this variable is updated from `Init`, which is called from `StartThread`. Is this thread safe?

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:18]: This comment looks obvious. Is this necessary?

File: plugins/obs-webrtc/whip-output.cpp
Comment (PatTheMav) [reply plugins/obs-webrtc/whip-output.cpp:175]: Correct would be either "macOS" or "Mac OS X", though the latter only identifies versions up to and including 10.15.

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [inline plugins/obs-webrtc/whip-output.cpp:230]: Please remove trailing spaces in the lines.

**Comment by PatTheMav** [comment]
macOS curl is 7.85 on CI runners (Xcode 14.2) and 7.87 on local machines (Xcode 14.3), so the version provided by the OS is recent enough to include the symbol.

Compiling the header API example file on CI and locally works without issue, so I have currently no clue what's the issue there.

**Comment by PatTheMav** [comment]
Found the issue - the cURL header declaring the new header API did not contain `extern "C"` for C++ compatibility until version 7.86. GitHub's macOS runners are stuck with Xcode 14.2 by default, which ships with the macOS 13.1 SDK.

The macOS 13.3 SDK contains the more recent version of cURL which fixes the issue for C++ projects, but the macOS 12 runners _do not_ provide Xcode 14.3 (only the macOS 13 runners do, but those do not support running AppleScripts yet, which we require to package the application).

So short of manually installing Xcode 14.3 on the runners, this will not succeed on macOS until GitHub either provides Xcode 14.3 on macOS 12 runners _or_ updates macOS 13 runners to also run AppleScript.

File: plugins/obs-webrtc/whip-output.h
Comment (tt2468) [reply plugins/obs-webrtc/whip-output.h:27]: I believe `with a lower value` would be a more correct way of phrasing it.

**Comment by PatTheMav** [comment]
I've authored https://github.com/obsproject/obs-studio/pull/9143 which should fix the build issues once merged.

File: plugins/obs-webrtc/whip-output.cpp
Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:228]: @norihiro is this something to watch out for, I assumed clang-format handled this scenario?

File: plugins/obs-webrtc/whip-output.h
Comment (mondain) [reply plugins/obs-webrtc/whip-output.h:27]: Darn it, failing at my native language :)

File: plugins/obs-webrtc/whip-output.cpp
Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:18]: I'd say no; sometimes I forget to remove "notes to self"

File: plugins/obs-webrtc/whip-output.cpp
Comment (pkviet) [reply plugins/obs-webrtc/whip-output.cpp:175]: i get this for safari UA :
`Mozilla/5.0 (Macintosh; Intel Mac OS X 13_4_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Safari/605.1.15`
(the version you quote is from April 2014 😉 )

File: plugins/obs-webrtc/whip-output.cpp
Comment (pkviet) [reply plugins/obs-webrtc/whip-output.cpp:173]: we're dropping support for windows x86 so any reference to arch can be dropped actually

Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:173]: @pkviet excellent; I'll remove it before my next push for clearing up these latest review items.

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [reply plugins/obs-webrtc/whip-output.cpp:18]: If I understand correctly, `StartThread` calls `Init` at first so that the variable `user_agent` is updated every time the WHIP output started.

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [reply plugins/obs-webrtc/whip-output.cpp:168]: I cannot find such a usage that the operator `<<` is used only once for a `std::stringstream` variable.

Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:168]: @norihiro right, but let's not forget your mod which I pulled in :)  I'll get it changed

File: plugins/obs-webrtc/whip-output.cpp
Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:18]: @norihiro true, what do you suggest as far as making it constant, or am I misunderstanding?

Comment (norihiro) [reply plugins/obs-webrtc/whip-output.cpp:18]: If I were you, I would put `user_agent` as a member variable of the class `WHIPOutput` or construct the user agent everytime the user agent is required because it is required only 3 times for each pair of start-stop.
Thinking 3rd party plugins like obs-multi-rtmp and the possibility of more features in obs, in my opinion, the output should be designed to be thread-safe.

**Comment by murillo128** [comment]
Note that using OPTIONS for turn server config is not the preferred method for WHIP:

> However, this method is not NOT RECOMMENDED and if supported by the underlying WHIP Client's webrtc implementation, the WHIP Client SHOULD wait for the information to be returned by the WHIP Endpoint on the response of the HTTP POST request instead.

It would be great if we could add support for the `setConfiguration` method in the libdatachannel peerconneciton object instead to add the turn server info received on the POST response.

**Comment by murillo128** [comment]
The OPTIONS requests only provides the stun/turn server info, not the ICE candidates. You should not wait for the STUN/TURN server gathering/allocation before sending the POST request to the WHIP endpoint.

So you will have to send the PATCH request anyway regardless if you have retrieved the stun/turn server info via POST or OPTIONS.

**Comment by mondain** [comment]
@murillo128 Yes, well aware of what comes in the OPTIONS. What happens in libdatachannel (not affiliated with OBS studio) with the iceServers information and process is in the library, not OBS. No PATCH is currently required when OPTIONS is used and its TURN server information is passed to iceServers in the context for libdatachannel. This process is very similar if not identical to how browsers work when provided with iceServers.

**Comment by murillo128** [comment]
The ice gathering is an async operation, which according to the spec is inited when the remote description is set (chrome deviates from the spec and starts gathering when the PC is created).

Anyway, if you receive the STUN/TURN server info, and then create the peerconnection and generate the sdp offer, it won't have any stun/turn ICE candidates available on it. 

You will have to wait for libdatachannel to fire the `pc.onLocalCandidate()` callback and send it back on a PATCH request:
https://github.com/paullouisageneau/libdatachannel#signal-a-peerconnection

**Comment by mondain** [comment]
@murillo128 I'll have to verify the libdatachannel process flow for that particular item, but for now its outside the scope of this PR; a PATCH process could most certainly increase the capability and value for the obs-webrtc plugin to be sure. As an aside, all browsers our team tested with for WHIP all play well with OPTIONS and provide srflx candidates in their offer SDP. There may be some magic that our front-end team did, but I'd imagine it was primarily a function of the browser itself (Not just Chrome).

**Comment by mondain** [comment]
@murillo128 yeah, don't get me wrong, I prefer to hard-code nothing :)

**Comment by murillo128** [comment]
I have started to implement the required changes to support STUN/TURN servers on the POST response here:

https://github.com/paullouisageneau/libdatachannel/pull/920

File: plugins/obs-webrtc/whip-output.cpp
Comment (murillo128) [inline plugins/obs-webrtc/whip-output.cpp:13]: we need to make sure that `video_ssrc` is not the same as `audio_ssrc`

Comment (mondain) [reply plugins/obs-webrtc/whip-output.cpp:13]: Its not

**Review by tt2468 — APPROVED** [review]

**Comment by mondain** [comment]
@tt2468 if it gets us finally merged for such a minor PR yields a more capable WebRTC feature, I'm fine with "predictable" SSRC's; although without looking it up in the RFC's and locating citations, I'm fairly certain this is not any expectation of SSRC's. No browser that I'm aware of produces predictable SSRC's.

**Review by RytoEX — CHANGES_REQUESTED** [review]
The commits need squashed before merging, and the final commit(s) will need to be appropriately prefixed with "obs-webrtc: ".

File: plugins/obs-webrtc/whip-output.cpp
Comment (murillo128) [reply plugins/obs-webrtc/whip-output.cpp]: there is no warranty that the values returned by the function in two consecutive invocations can't be equal. The probability is low (very low), but nevertheless possible.

**Comment by mondain** [comment]
I'm not aware of a requirement to have atomic decay level randomness on a
single instance on a single computer doing media streams with 1 audio and 1
video; there's good enough and there's nitpicking for no good purpose.

File: plugins/obs-webrtc/whip-output.cpp
Comment (DDRBoxman) [inline plugins/obs-webrtc/whip-output.cpp:96]: Is this variable necessary? I guess it helps with readability a bit when you look at the companion video function.

File: plugins/obs-webrtc/whip-output.h
Comment (DDRBoxman) [inline plugins/obs-webrtc/whip-output.h:123]: Can we pull the below functions out into a util file?

**Review by DDRBoxman — APPROVED** [review]
Looks pretty solid, left some nitpicks. 👍

File: plugins/obs-webrtc/whip-output.cpp
Comment (norihiro) [reply plugins/obs-webrtc/whip-output.cpp]: If we have multiple output from obs-webrtc, eg. stop and start again or start multiple outputs by a future change, the same SSRCs will be used among multiple outputs. Is that ok?
If not, `video_ssrc` and `audio_ssrc` should be generated for each instance and stored in the class `WHIPOutput`.

**Comment by murillo128** [comment]
> I'm not aware of a requirement to have atomic decay level randomness on a
single instance on a single computer doing media streams with 1 audio and 1
video; there's good enough and there's nitpicking for no good purpose.

**Comment by murillo128** [comment]
>I'm not aware of a requirement to have atomic decay level randomness on a
single instance on a single computer doing media streams with 1 audio and 1
video; there's good enough and there's nitpicking for no good purpose.

I assume this is about the ssrc geenreation. There is a hard requirement that the audio and video ssrc must be different, and generating two random values is not going to ensure that. 

I think I saw a proposal of just using `audio = random()` and `video = audio+1` that will trivially solve that. We would also want to avoid having an `ssrc = 0` as I have seen some lazy implementations failing in that case.

File: plugins/obs-webrtc/whip-output.cpp
Comment (PatTheMav) [inline plugins/obs-webrtc/whip-output.cpp:26]: As I'm not familiar with WebRTC, can you explain to me how this change is related to user agents and `OPTIONS` requests?

While the PR has been dutifully squashed into a single commit, its message will not suggest to me that `ssrc` generation was also changed as part of it. I'm not opposed to merging it as part of this PR, but I'd rather see those changes moved into its own commit (so _if_ issues arise with it, we don't have to manually pull the changes apart to revert it).

**Comment by PatTheMav** [comment]
> > I'm not aware of a requirement to have atomic decay level randomness on a
> > single instance on a single computer doing media streams with 1 audio and 1
> > video; there's good enough and there's nitpicking for no good purpose.
> 
> I assume this is about the ssrc geenreation. There is a hard requirement that the audio and video ssrc must be different, and generating two random values is not going to ensure that.
> 
> I think I saw a proposal of just using `audio = random()` and `video = audio+1` that will trivially solve that. We would also want to avoid having an `ssrc = 0` as I have seen some lazy implementations failing in that case.

As implemented, at both video or audio `ssrc` can be zero (either because the randomizer yields `0` as the video `ssrc` or we overflow `UINT32_MAX` to `0` by adding `1`).

Using `std::uniform_int_distribution<uint32_t> dist(1, (UINT32_MAX-1));` should fix, as it should neither yield `0` itself, nor a value that would overflow to `0` (iirc `dist` uses inclusive boundaries).

**Review by DDRBoxman — CHANGES_REQUESTED** [review]
Can you fix up the title and description on this PR?

Looks like it's just user agent, and timestamp / ssrc generation now.

**Review by DDRBoxman — APPROVED** [review]

**Review by DDRBoxman — CHANGES_REQUESTED** [review]
See comment above: https://github.com/obsproject/obs-studio/pull/9126#issuecomment-1614543613

File: plugins/obs-webrtc/whip-output.h
Comment (PatTheMav) [inline plugins/obs-webrtc/whip-output.h:30]: I'd prefer if this were a `static int` (there is no need to use a preprocessor define for this), but won't hold up merging if this isn't changed.

File: plugins/obs-webrtc/whip-output.cpp
Comment (tt2468) [reply plugins/obs-webrtc/whip-output.cpp:96]: It's for code readability mainly, yes. I didn't want there to be confusion about the method being used for SSRC creation. As you can see in the video track function, it's the same style.

**Review by DDRBoxman — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]
Merging based on approvals from @DDRBoxman and @tt2468 .

**Pull request merged**
